nixos/modules/services/development/guix.nix

```diff
@@ -0,0 +1,118 @@
+{config, pkgs, lib, ...}:
+
+with lib;
+
+let
+
+  cfg = config.services.guix;
+
+  guix-binary = pkgs.stdenv.mkDerivation rec
+    { name = "guix-binary-${version}";
+      version = "0.16.0";
+
+      src = pkgs.fetchurl {
+        # TODO: make work on non x86_64 systems
+        url = "https://alpha.gnu.org/gnu/guix/guix-binary-${version}.x86_64-linux.tar.xz";
+        sha256 = "049l0zim30cd0gyly2h3jaw4cshdk78h7xdb9ac173h72i13afbj";
+      };
+      sourceRoot = ".";
+
+      outputs = [ "out" "store" "var" ];
+      phases = [ "unpackPhase" "installPhase" ];
+
+      installPhase = ''
+        # copy the /gnu/store content
+        mkdir -p $store
+        cp -r gnu $store
+
+        # copy /var content
+        mkdir -p $var
+        cp -r var $var
+
+        # link guix binaries
+        mkdir -p $out/bin
+        ln -s /var/guix/profiles/per-user/root/current-guix/bin/guix $out/bin/guix
+        ln -s /var/guix/profiles/per-user/root/current-guix/bin/guix-daemon $out/bin/guix-daemon
+      '';
+    };
+
+  buildGuixUser = i:
+    {
+      "guixbuilder${builtins.toString i}" = {
+        group = "guixbuild";
+        extraGroups = ["guixbuild"];
+        home = "/var/empty";
+        shell = pkgs.nologin;
+        description = "Guix build user ${builtins.toString i}";
+        isSystemUser = true;
+      };
+    };
+
+in {
+
+  options.services.guix = {
+    enable = mkEnableOption "GNU Guix package manager";
+  };
+
+  config = mkIf (cfg.enable) {
+
+    users = {
+      extraUsers = lib.fold (a: b: a // b) {} (builtins.map buildGuixUser (lib.range 1 10));
+      extraGroups.guixbuild = {name = "guixbuild";};
+    };
+
+    environment.systemPackages = [ guix-binary ];
+
+    systemd.services.guix-daemon = {
+      enable = true;
+      description = "Build daemon for GNU Guix";
+      serviceConfig = {
+        ExecStart="/var/guix/profiles/per-user/root/current-guix/bin/guix-daemon --build-users-group=guixbuild";
+        Environment="GUIX_LOCPATH=/var/guix/profiles/per-user/root/guix-profile/lib/locale";
+        RemainAfterExit="yes";
+        StandardOutput="syslog";
+        StandardError="syslog";
+        TaskMax= "8192";
+      };
+      wantedBy = [ "multi-user.target" ];
+    };
+
+    system.activationScripts.guix = ''
+
+      # copy initial /gnu/store
+      if [ ! -d /gnu/store ]
```

Comment:
This breaks when updating`guix-binary`. Its new packaged `/gnu/store` doesn't get copied.
I think merging them wouldn't be harmful because the store paths don't change. I.e. we could always rsync from the package to the gnu-store. I also don't expect the daemon to be restarted a lot.

Maybe someone else has a better idea.

Reply:
Hm. This might be tricky.

I think merging would be a non-op. The [Guix Store database](https://www.gnu.org/software/guix/manual/en/html_node/The-Store.html) (located in `/var/guix/db`) would not be updated with new stuff added to `/gnu/store` and guix would probably garbage collect it away. The database would need to be updated, but that seems a bit far out.

Guix can install an update of the guix binary into the `/gnu/store` and will then use that. Weird things might happen when guix in `/gnu/store` is of a different version than guix in `/nix/store`.

Reply:
Well do we want to put guix in the nix-store at all?
Maybe just add the guix in `/gnu/store` to the path.

Reply:
Works! Even recommended by the [Guix manual](https://www.gnu.org/software/guix/manual/en/html_node/Binary-Installation.html#Binary-Installation):

>  Make the guix command available to other users on the machine, for instance with:
> 
> # mkdir -p /usr/local/bin
> # cd /usr/local/bin
> # ln -s /var/guix/profiles/per-user/root/current-guix/bin/guix

See: https://github.com/NixOS/nixpkgs/pull/56430/commits/d95fa0394fdb606c8b33b5a58d862aa286d4902c